x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -338,82 +351,133 @@ public void removeFromCache(final CacheKey cacheKey) {
-     * while the current thread is holding the lock associated to the shard.
+     * Waits for the cache files associated with the shard represented by ({@link SnapshotId}, {@link IndexId}, {@link SnapshotId}) to be
+     * evicted if the shard is marked as evicted in cache at the time this method is executed.
      *
-     * @param snapshotId the snapshot the evicted searchable snapshots shard belongs to
-     * @param indexId    the index in the snapshot the evicted searchable snapshots shard belongs to
-     * @param shardId    the searchable snapshots shard id
-     * @param runnable   a runnable to execute
+     * @param snapshotUUID      the snapshot's unique identifier
+     * @param snapshotIndexName the name of the index in the snapshot
+     * @param shardId           the {@link ShardId}
      */
-    public void runIfShardMarkedAsEvictedInCache(SnapshotId snapshotId, IndexId indexId, ShardId shardId, Runnable runnable) {
-        runIfShardMarkedAsEvictedInCache(new ShardEviction(snapshotId.getUUID(), indexId.getName(), shardId), runnable);
+    public void waitForCacheFilesEvictionIfNeeded(String snapshotUUID, String snapshotIndexName, ShardId shardId) {
+        final Future<?> future;
+        synchronized (shardsEvictionsMutex) {
+            if (allowShardsEvictions == false) {
+                throw new AlreadyClosedException("Cannot wait for shard eviction to be processed, cache is stopping");
+            }
+            future = pendingShardsEvictions.get(new ShardEviction(snapshotUUID, snapshotIndexName, shardId));
+            if (future == null) {
+                return;
+            }
+        }
+        FutureUtils.get(future);
     }
 
     /**
-     * Allows to run the specified {@link Runnable} if the shard represented by {@link ShardEviction} is still marked as evicted at the time
-     * this method is executed. The @link Runnable} will be executed while the current thread is holding the lock associated to the shard.
+     * Evicts the cache files associated to the specified {@link ShardEviction}.
      *
-     * @param shardEviction a {@link ShardEviction} representing the shard marked as evicted
-     * @param runnable      a runnable to execute
+     * @param shardEviction the shard eviction to process
      */
-    private void runIfShardMarkedAsEvictedInCache(ShardEviction shardEviction, Runnable runnable) {
-        try (Releasable ignored = shardsEvictionLock.acquire(shardEviction)) {
-            boolean success = false;
+    void processShardEviction(ShardEviction shardEviction) {
+        assert isPendingShardEviction(shardEviction) : "shard is not marked as evicted: " + shardEviction;
+        assert assertGenericThreadPool();
+
+        shardsEvictionsLock.readLock().lock();
+        try {
             try {
-                if (evictedShards.remove(shardEviction)) {
-                    runnable.run();
+                if (allowShardsEvictions) {
```

Comment:
I think (but did not check) that the JMM allows this read to be OK, since we lock the write lock after setting the value, but given that this is not performance sensitive or anything, I would prefer to protect the read of the boolean by grabbing the mutex around reading it.

Reply:
Thanks for catching this, I pushed ba8a029447efc01f5ad589017954e536e8a815b3.